### ISSUE TITLE
`null is T` does not promote `null` to `T`

### ISSUE THREAD

**Issue by ds84182** [dart-lang/language#3116]
The following works:

```dart
T find<T extends Object?>() {
  /* ... loop with early return here ... */
  const nil = null;
  if (nil is T) return nil; // Promotes const nil to T
  throw StateError('Not found');
}
```

But this doesn't:

```dart
T find<T extends Object?>() {
  /* ... loop with early return here ... */
  if (null is T) return null; // Does not promote null to T
  throw StateError('Not found');
}
```

Kinda unexpected that consts can be promoted though. That's cool.

EDIT: This is not actually due to promotion of local consts. See https://github.com/dart-lang/language/issues/3116#issuecomment-1566232588

**Comment by mateusfccp** [comment]
I think this is expected. `null` is not an identifier, it's a literal.

**Comment by ds84182** [comment]
I disagree. Either both scenarios should fail, or both should pass. Type promotion is only specified to work on local variables.

Constant declarations within a function are not "variable" but instead are aliases for the referred constant and should be semantically the same as replacing the use sites with the underlying constant value as part of the declaration's initializer.

I'd expect this reflexivity to be upheld but it isn't due to type promotion of the constant declaration's type. Literals should be considered the same as their aliases unless bound to a local or static variable.

**Comment by lrhn** [comment]
> I disagree.

You're free to disagree. It's still true, though. This _is_ expected behavior, because you can only promote variables, and `null` is not a variable.

> This teeters the edge between a spec bug or an implementation bug.

Not even close. You can claim it's a missing feature, but it's not a bug, because it is working as both intended and specified. 

There is no support in the language specification for "promoting" a non-variable, or applying a lower bound to a type variable, either of which could make a test like `null is T` allow a latter occurrence of `null` to be known to be a subtype of `T`.


So, what would it take to make this work?

* One option is to allow *constant expressions* to be treated like variables. We know, at compile-time, whether two constant expressions refer to the *same* value, and it's definitely not changing, so any `c is T` check could promote `c` to `T` the same way it could a variable, and later occurrences of the same variable could use the promoted type. 
That only really makes sense if the type being tested against includes a type variable.

* Another option is promoting *type variables*, where we gather information about the type represented by a type variable by doing tests *with* it. That's generally *lower-bound* information, so `null is T` would tell us that `T` is a supertype of `Null`, and `1 is T` would tell us that `T` is a supertype of `int`.  We don't have lower-bounds elsewhere, so it's more tricky to introduce, but if it's something we can erase at runtime, it might be possible.
*

**Comment by ds84182** [comment]
> Not even close. You can claim it's a missing feature, but it's not a bug, because it is working as both intended and specified.

When I looked at the specification, it only said promotion happens to local variables. It's unclear whether this means it applies to local consts, and I assumed it did not & should not.

However I now realize the code sample actually isn't promotion. The const is typed as `dynamic`, giving it an explicit type of `Null` shows there _isn't_ any promotion after all.

**Comment by lrhn** [comment]
There is promotion for constant local variables too. The simple example is:
```dart
void main() {
  const Object x = "A";
  if (x is String) print(x.codeUnits); // [65]
}
```
This shows promotion from `Object` to `String` of the local (constant) variable `x`.

There does appear to be a bug with promotion of a `const` variable using a type variable, though. I'll file an issue for that.
```dart
void main<T>() {
  const Object? x = "a";
  if (x is T) {
    T value = x; // Error, but none if `const` is changed to `final`
  }
}
```

**Comment by ds84182** [comment]
> The catch is that you can only promote to a subtype, and an unbounded type variable is not a subtype of anything but top types.

Ah, that's what makes this confusing. The true static type has to be hidden from the type system for promotion to work. It guards against nonsensical promotions (e.g. `if (string is int) string + 123;`) but also prevents seemingly fine promotions from succeeding.

Type promotion would need to be smarter for either promotion option to work.

> * Another option is promoting _type variables_, where we gather information about the type represented by a type variable by doing tests _with_ it.

I think this option would be somewhat nicer to have. But type variable promotion via `is` check is pretty opaque. A type test like `if (T extends Null)` would be easier to understand.

**Comment by rrousselGit** [comment]
One a different topic, I could see `null` be represented as a variable:

```dart
const null = Null();
```

**Comment by lrhn** [comment]
Even if we declared a variable `const Null null = Null._();`, where we'd have a magical `external const factory Null._();`, it wouldn't promote because we only promote local variables.

We could choose to promote static `const` variables too. We don't generally promote `final` variables, because they might be changed to a getter in a later version, and we want that to be non-breaking, but you can't change a `const` variable to a getter, that's already breaking.

But then we might as well promote any constant *value*, like I suggested above. A constant variable is just one way to refer to the same constant value.

**Comment by munificent** [comment]
I'm going to go ahead and close this because it's working as intended.
